core/src/main/java/org/elasticsearch/action/ActionRegistrar.java

```diff
@@ -0,0 +1,388 @@
+import org.elasticsearch.action.ingest.PutPipelineAction;
+import org.elasticsearch.action.ingest.PutPipelineTransportAction;
+import org.elasticsearch.action.ingest.SimulatePipelineAction;
+import org.elasticsearch.action.ingest.SimulatePipelineTransportAction;
+import org.elasticsearch.action.main.MainAction;
+import org.elasticsearch.action.main.TransportMainAction;
+import org.elasticsearch.action.search.ClearScrollAction;
+import org.elasticsearch.action.search.MultiSearchAction;
+import org.elasticsearch.action.search.SearchAction;
+import org.elasticsearch.action.search.SearchScrollAction;
+import org.elasticsearch.action.search.TransportClearScrollAction;
+import org.elasticsearch.action.search.TransportMultiSearchAction;
+import org.elasticsearch.action.search.TransportSearchAction;
+import org.elasticsearch.action.search.TransportSearchScrollAction;
+import org.elasticsearch.action.support.TransportAction;
+import org.elasticsearch.action.termvectors.MultiTermVectorsAction;
+import org.elasticsearch.action.termvectors.TermVectorsAction;
+import org.elasticsearch.action.termvectors.TransportMultiTermVectorsAction;
+import org.elasticsearch.action.termvectors.TransportShardMultiTermsVectorAction;
+import org.elasticsearch.action.termvectors.TransportTermVectorsAction;
+import org.elasticsearch.action.update.TransportUpdateAction;
+import org.elasticsearch.action.update.UpdateAction;
+import org.elasticsearch.common.NamedRegistry;
+import org.elasticsearch.plugins.ActionPlugin;
+import org.elasticsearch.plugins.ClientActionPlugin;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Map;
+
+import static java.util.Collections.unmodifiableMap;
+
+/**
+ * {@link ActionRegistrar} is responsible for registering all of the core actions. It also allows for additional
+ * {@link ActionPlugin} and {@link ClientActionPlugin} to be passed in. The {@link ActionRegistrar} ensures all of the core
+ * actions are also in the {@link ActionRegistrar#getClientActions()}.
+ *
+ * Please note that the {@link org.elasticsearch.client.transport.TransportClient} uses the {@link ActionPlugin} objects
+ * to construct its routing table, and the {@link ClientActionPlugin} to construct its client side interactions.
+ * Ensure that they both contain the same {@link GenericAction}s.
+ */
+class ActionRegistrar {
+
+    private final ActionRegistry actions;
+    private final ClientActionRegistry clientActions;
+
+    Map<String, ActionPlugin.ActionHandler> getActions() {
+        return unmodifiableMap(actions.getRegistry());
+    }
+
+    Map<String, GenericAction<? extends ActionRequest, ? extends ActionResponse>> getClientActions() {
+        return unmodifiableMap(clientActions.getRegistry());
+    }
+
+    ActionRegistrar() {
+        this(Collections.emptyList(), Collections.emptyList());
+    }
+
+    ActionRegistrar(List<ActionPlugin> actionPlugins, List<ClientActionPlugin> clientActionPlugins) {
+        actions = new ActionRegistry();
```

Comment:
This should take the actions as ctor args, so that they can be final/unmodifiable.